Testing with Go VCR and stored fixtures

tbroden84 opened the pull request:
### WHY are these changes introduced?

During development to have the confidence to change, update, add new features, improvements and fixes. Cover the provider with tests as much as possible. For this Terraform acceptance test (extension of Go test) can be used. A downside to only relying on acceptance tests, is that all tests require infrastructure resources. Which takes time to create, update and clean up.

With the use of Go VCR, it's possible to record the communication to the API backend and store these as fixtures (cassettes). Then replay these instead of doing the real requests. This will speed up each test significantly.

Requires: https://github.com/84codes/go-api/pull/48 (merged)

### WHAT is this pull request doing?

- Adapt the Provider to be able to use Go VCR via http client when running acceptance test.
  - Record request/responses to fixtures
  - Replay requests/responses from fixtures
- Append resource configurations.
- Use templated configurations with dynamic values.
- Update Github CI workflow to run all tests on commit.

### HOW can this pull request be tested?

Run tests manually or through automated Github CI workflow.

tbroden84 commented:
Strange behaviour on `vpc_gcp_peering` test. During recording state it succeed but fails during playback. Says it's missing interaction for `/api/plans` and `/api/regions` used for online validation during `terraform plan` for instances. They are there in the fixture cassette in the initial requests. 

However seems to be an overall issue using two instances in one test. State-data are getting mixed up, first gets seconds attributes and vice verse.

tbroden84 commented:
Data sources and resources that covered with basic tests.
Data sources:
- alarm
- nodes
- notification

Resources:
- alarm
- aws_eventbridge
- extra_disk_size
- integration_log
- integration_metrics
- instance
- node_actions
- notification
- plugin
- plugin_community
- rabbitmq_configuration
- security_firewall
- vpc_connect
- webhook

tbroden84 commented:
Those that are skipped: 
Data sources:
- account_vpcs
- account
- credentials
- instance
- plugin_community
- plugin
- upgradable_version
- vpc_gcp_info
- vpc_info

Resources:
- account_actions
- custom_domain
- upgrade_rabbitmq
- vpc_gcp_peering (passes initial recording but fails during playback)
- vpc_peering

magnushoerberg reviewed (commented):
Needed to run `go mod tidy` so update needed for `go.mod` and `go.sum`?

I get this error:

`dotenv -f .env go test -v ./cloudamqp/`

```
=== RUN   TestAccIntegrationLog_Basic
    testing.go:705: Step 0 error: Check failed: 1 error occurred:
                * Check 9/23 error: cloudamqp_integration_log.cloudwatch: Attribute 'access_key_id' expected "", got "***"
```

magnushoerberg commented:
Test run is very fast now :) great work!

tbroden84 commented:
> Needed to run `go mod tidy` so update needed for `go.mod` and `go.sum`?

Will be solved once go-api is bumped with: https://github.com/84codes/go-api/pull/48

tbroden84 commented:
> I get this error:
> 
> `dotenv -f .env go test -v ./cloudamqp/`
> ...

Hmm, mine TestAccIntegrationLog_Basic test passes. Running 
`TF_ACC=1 dotenv -f .env go test -v ./cloudamqp/` otherwise the tests will be skipped.

Perhaps also need to share the .env for all tokens in log/metric integration.

tbroden84 commented on cloudamqp/resource_cloudamqp_extra_disk_size_test.go:
```diff
@@ -0,0 +1,113 @@
+	})
+}
+
+// Uncertain test and mostly fail locally due to backend cannot find the correct resource group
```
Had previous issues with finding correct resource group. Tried it once more and waited longer this time, to see if this could be solved by our backend. Found the resource group after a while, have now updated this test to be used together with its fixture.

tbroden84 commented:
@dentarg, @magnushoerberg after earlier discussion, all tests are active in the CI workflow. Also went over refactoring the sanitize of sensitive API key/token data (used by log & metrics integrations), re-recorded all tests and other smaller fixes.

Could anyone of you please just double check that it looks ok? Tried cover as much of the resources as possible with at least one basic test (create, import, delete) as a beginning. Then we can extend them over time. Would be nice to get this in place to further check the Terraform Plugin SDK v2 upgrade and move in go-api into the provider.

Notes:
Plugins and community plugins both tests takes 20 s each, due to sleeps. Tried to filter out the request we are waiting on, but couldn't find a good way to catch the correct data. All plugins will be fetched and filtered out based on the name of the plugin being enable. However don't have that information when building the `shouldSaveHook`.

Webhook, re-recording keeps on failing due to timeout in the backend. Will have to update go-api to handle this, together that we also support updates. See this as a separated PR specific for this with a new recording.

dentarg commented:
> Plugins and community plugins both tests takes 20 s each, due to sleeps.

We could make this configurable for users, and use it ourselves in the tests? (could be a follow-up PR if so)

tbroden84 commented:
Yeah, while re-recording and without filter. We still have requests made between sleep, this halved the time playing the test. Used old recorded cassette with lower sleep in replay lower in down to about 2 s instead. Updated the sleep and added a comment about using default sleep during record to both tests.

dentarg commented:
I see, when recording with a very small sleep value, there will be many requests made, not optimal. Nice with the comment for that. Could possible be configured via `ENV`? So there is one set of settings for recording, another set for running the tests. Something we can explore later on.

Can we do something about `TestAccVpcConnect_Azure_Basic`? Takes 10s

tbroden84 commented:
> Could possible be configured via `ENV`?

Good point.

> Can we do something about TestAccVpcConnect_Azure_Basic? Takes 10s

Will check.

tbroden84 commented:
New basic test for webhook resource, it has been updated in the provider. In order to test updating the webhook resource, the queue must exists. Otherwise the validation and state file will contain wrong information after the update. Would require a secondary provider, that can access message broker HTTP API.

dentarg commented:
How did you solve that? Did you had a queue by hand before recording?

tbroden84 commented:
Async operation, so the initial create of the webhook will return "correct" value. Then update the status once all concurrent tries have been failing. While doing the update, this is not the case.

Would preferable be made with the queues, so the real scenario can be tested. Including update.

tbroden84 commented:
Never did a deep dive into the webhook application. Looks like we have an issue in the API handling.

dentarg commented:
> Async operation, so the initial create of the webhook will return "correct" value.

Sounds good enough for this PR?

tbroden84 commented:
> > Async operation, so the initial create of the webhook will return "correct" value.
> 
> Sounds good enough for this PR?

Yes I would say so, similar to other integrations tests that making sure it can be created in the API. Did some investigation, used the wrong request body earlier today while double checking the API. Didn't make sense that the test should fail with update. Made adjustment and re-recording to also include the update step: https://github.com/cloudamqp/terraform-provider-cloudamqp/pull/257/commits/985c125d16c8ffa6a0408186f1bd3fb295e8c8d1

dentarg reviewed (approved):
Awesome work here :) 🚀 📻
